torch/csrc/autograd/autograd.cpp

```diff
@@ -0,0 +1,129 @@
+            "tuple is not a Tensor", i);
+        const Variable& variable = as_variable_ref(output);
+        auto gradient_edge = variable.gradient_edge();
+        TORCH_CHECK(gradient_edge.function,
+            "element %d of tensors does not require grad and does not have a grad_fn", i);
+        roots.push_back(std::move(gradient_edge));
+
+        at::Tensor grad = grad_outputs[i];
+        if (!grad.defined()) {
+            // if grad is undefined
+            if (output.requires_grad()) {
+                TORCH_CHECK(output.numel() == 1, "grad can be implicitly created only for scalar outputs");
+                // create the grad with ones tensor for output that requires grad
+                grads.push_back(at::ones_like(output));
+            }
+            // otherwise we don't put grad into the grads list
+        } else if (grad.is_variable()) {
+            // if grad is a typical variable
+            // grads.push_back(as_variable_ref(grad));
+            grads.push_back(as_variable_ref(grad));
+            std::cout<<"grad :" << grads[0] << std::endl;
+        } else {
+            AT_ERROR("element %d of gradients tuple is not a Tensor or None", i);
+        }
+    }
+
+    edge_list output_edges;
+    if (inputs.empty()) {
+        size_t num_inputs = inputs.size();
+        output_edges.reserve(num_inputs);
+        for (size_t i = 0; i < num_inputs; ++ i) {
+            const Variable& input = as_variable_ref(inputs[i]);
+            const auto output_nr = input.output_nr();
+            auto grad_fn = input.grad_fn();
+            if (!grad_fn) {
+                grad_fn = input.try_get_grad_accumulator();
+            }
+            TORCH_CHECK(input.requires_grad(),
+                "One of the differentiated Tensors does not require grad");
+            if (!grad_fn) {
+                output_edges.emplace_back();
+            } else {
+                output_edges.emplace_back(grad_fn, output_nr);
+            }
+        }
+    }
+    std::cout<<"roots size: " << roots.size() << " grads size: " << grads.size() << std::endl;
+
+    return Engine::get_default_engine().execute(roots, grads, keep_graph, create_graph, output_edges);
+   }
+
+  void backward(
+      at::TensorList tensors,
+      at::TensorList grad_tensors,
+      c10::optional<bool> keep_graph,
+      bool create_graph) {
+    if (!keep_graph) {
+        keep_graph = create_graph;
+    }
+    if (grad_tensors.empty()) {
+        std::vector<at::Tensor> new_grads;
+        for(size_t i = 0; i < tensors.size(); ++ i) {
+            new_grads.emplace_back(at::Tensor());
+        }
+        grad_tensors = new_grads;
+    }
+    run_backward(tensors, grad_tensors, keep_graph.value(), create_graph, {});
+  }
+
+  variable_list grad(
+      at::TensorList outputs,
+      at::TensorList inputs,
+      at::TensorList grad_outputs,
+      c10::optional<bool> keep_graph,
+      bool create_graph,
+      bool allow_unused) {
```

Comment:
Ditto for torch.autograd.grad signature: `torch.autograd.grad(outputs, inputs, grad_outputs=None, retain_graph=None, create_graph=False, only_inputs=True, allow_unused=False)`

Reply:
same for `only_inputs`, it's explicitly deprecated and it's a no-op in python, I think C++ API can just ignore this? Other than the deprecated flags, I think we matched the `python` semantics/signatures in `autograd.h`.